lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -462,110 +578,138 @@ class DiscoverLoops {
-                 bool baseCase = true) {
-    auto refersToIter = valRefersTo.find(val);
-    if (refersToIter != valRefersTo.end()) {
-      for (auto ref : refersToIter->second)
-        if (f(ref).failed())
-          return;
-    } else if (baseCase) {
-      FieldRef base(val, 0);
-      if (f(base).failed())
-        return;
-    }
+  // Add both the probes to the same equivalence class, to record that they
+  // refer to the same data value.
+  void probesReferToSameData(Value probe1, Value probe2) {
+    auto p1Node = getOrAddNode(probe1);
+    auto p2Node = getOrAddNode(probe2);
+    rwProbeClasses.unionSets(p1Node, p2Node);
   }
 
-  void dump() {
-    for (const auto &valRef : valRefersTo) {
-      llvm::dbgs() << "\n val :" << valRef.first;
-      for (auto node : valRef.second)
-        llvm::dbgs() << "\n Refers to :" << getFieldName(node).first;
-    }
-    for (const auto &dtv : fieldToVals) {
-      llvm::dbgs() << "\n Field :" << getFieldName(dtv.first).first
-                   << " ::" << dtv.first.getValue();
-      for (auto val : dtv.second)
-        llvm::dbgs() << "\n val :" << val;
-    }
-    for (const auto &p : portPaths) {
-      llvm::dbgs() << "\n Output port : " << getFieldName(p.first).first
-                   << " has comb path from :";
-      for (const auto &src : p.second)
-        llvm::dbgs() << "\n Input port : " << getFieldName(src).first;
-    }
+  void addToPortPathsIfRWProbe(unsigned srcNode,
+                               DenseSet<FieldRef> &inputPortPaths) {
+    // Check if there exists any RWProbe for the srcNode.
+    auto baseFieldRef = reachingDefGraph[srcNode].first;
+    if (auto defOp = dyn_cast_or_null<Forceable>(baseFieldRef.getDefiningOp()))
+      if (defOp.isForceable() && !defOp.getDataRef().use_empty()) {
+        // Assumption, the probe must exist in the equivalence classes.
+        auto rwProbeNode =
+            rwProbeClasses.getLeaderValue(getOrAddNode(defOp.getDataRef()));
+        // For all the probes, that are in the same eqv class, i.e., refer to
+        // the same value.
+        for (auto probe :
+             llvm::make_range(rwProbeClasses.member_begin(
+                                  rwProbeClasses.findValue(rwProbeNode)),
+                              rwProbeClasses.member_end())) {
+          auto probeVal = reachingDefGraph[probe].first;
+          // If the probe is a port, then record the path from the probe to the
+          // input port.
+          if (probeVal.getValue().isa<BlockArgument>()) {
+            inputPortPaths.insert(probeVal);
+          }
+        }
+      }
   }
 
+private:
   FModuleOp module;
   InstanceGraph &instanceGraph;
-  /// Map of a Value to all the FieldRefs that it refers to.
-  DenseMap<Value, SetOfFieldRefs> valRefersTo;
 
-  DenseMap<Value, DenseSet<Value>> aliasingValuesMap;
-
-  DenseMap<FieldRef, DenseSet<Value>> fieldToVals;
+  /// Map of value to the corresponding FieldRef.
```

Comment:
Words are hard, but is this something like: "Map of values to the set of all FieldRefs (same base) that this may be directly derived from through indexing operations."? Would help in understanding the various data-structures involved in this pass.

Reply:
Thanks, updated it.